File: Base de Dtos/turismo-spring-boot/src/main/java/pe/edu/upeu/turismospringboot/model/entity/LugarTuristico.java
package pe.edu.upeu.turismospringboot.model.entity;

import com.fasterxml.jackson.annotation.JsonBackReference;
import com.fasterxml.jackson.annotation.JsonManagedReference;
import jakarta.persistence.*;
import lombok.Data;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "lugar_turistico")
@Data
public class LugarTuristico {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long idLugar;

    private String nombre;
    private String descripcion;
    private String ubicacion;
    private String imagenUrl;

    @ManyToOne
    @JoinColumn(name = "id_categoria", nullable = false)
    @JsonBackReference
    private Categoria categoria;

    @OneToMany(mappedBy = "lugar", cascade = CascadeType.ALL)
    @JsonManagedReference
    private List<Emprendimiento> emprendimientos = new ArrayList<>();

    @OneToMany(mappedBy = "lugar", cascade = CascadeType.ALL)
    @JsonManagedReference
    private List<Resena> resenas = new ArrayList<>();

    private LocalDateTime fechaCreacionLugarTuristico;
    private LocalDateTime fechaModificacionLugarTuristico;

    @PrePersist
    public void onCreate(){
        fechaCreacionLugarTuristico = LocalDateTime.now();
    }

    @PreUpdate
    public void onUpdate(){
        fechaModificacionLugarTuristico = LocalDateTime.now();
    }
}

File: ProyectoTurismo/app/src/main/java/pe/edu/upeu/proyecto_turismo/modelo/LugarTuristico.kt
package pe.edu.upeu.proyecto_turismo.modelo

data class LugarTuristicoDto(
    val idLugar: Long = 0,
    val descripcion: String = "",
    val fechaCreacionLugarTuristico: String = "",
    val fechaModificacionLugarTuristico: String = "",
    val imagenUrl: String = "",
    val nombre: String = "",
    val ubicacion: String = "",
    val idCategoria: Long = 0
)

data class LugarTuristicoResp(
    val idLugar: Long,
    val descripcion: String,
    val fechaCreacionLugarTuristico: String,
    val fechaModificacionLugarTuristico: String,
    val imagenUrl: String,
    val nombre: String,
    val ubicacion: Ubicacion,
    val categoria: Categoria
)

fun LugarTuristicoResp.toDto(): LugarTuristicoDto {
    return LugarTuristicoDto(
        idLugar = this.idLugar,
        descripcion = this.descripcion,
        fechaCreacionLugarTuristico = this.fechaCreacionLugarTuristico,
        fechaModificacionLugarTuristico = this.fechaModificacionLugarTuristico,
        imagenUrl = this.imagenUrl,
        nombre = this.nombre,
        ubicacion = this.ubicacion.toString(),
        idCategoria = this.categoria.idCategoria
    )
}


